### PR TITLE
Adding support to mimetype autodetection

### PR DESCRIPTION
**Pull request by deltakosh**
This feature adds support from automatic mime type discovery which was missing

I was reluctant at first (#15793) because it introduces a breaking change.

That being said it is really unlikely that someone would make any use of what was returned.
I also started to add support for fetch (indirectly) as this is something we want to do.

Finally I flagged the non async functions as deprecated to help move user codebase

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Loading/sceneLoader.ts [hunk 1]
```diff
@@ -582,8 +608,8 @@ function loadData(
         // For plugin factories, the plugin is instantiated on each SceneLoader operation. This makes options handling
         // much simpler as we can just pass the options to the factory, rather than passing options through to every possible
         // plugin call. Given this, options are only supported for plugins that provide a factory function.
-        if (isFactory(registeredPlugin.plugin)) {
-            const pluginFactory = registeredPlugin.plugin;
+        if (isFactory(registeredPlugin!.plugin)) {
+            const pluginFactory = registeredPlugin!.plugin;
```
Comment (ryantrem) [inline packages/dev/core/src/Loading/sceneLoader.ts:612]: These non-null assertions should not be needed, correct?

Comment (deltakosh) [reply packages/dev/core/src/Loading/sceneLoader.ts:612]: Typescript disagreed with that (I was like you ;))

File: packages/dev/core/src/Loading/sceneLoader.ts [hunk 2]
```diff
@@ -777,13 +804,13 @@ function importMesh(
     pluginExtension: Nullable<string> = null,
     name = "",
     pluginOptions: PluginOptions = {}
-): Nullable<ISceneLoaderPlugin | ISceneLoaderPluginAsync> {
+): Promise<Nullable<ISceneLoaderPlugin | ISceneLoaderPluginAsync>> {
```
Comment (ryantrem) [inline packages/dev/core/src/Loading/sceneLoader.ts:807]: Why change this? It doesn't seem like there is any new asynchrony here, and it introduces a breaking change.

File: packages/dev/core/src/Loading/sceneLoader.ts [hunk 3]
```diff
@@ -1520,7 +1547,7 @@ export class SceneLoader {
      * @param onError a callback with the scene, a message, and possibly an exception when import fails
      * @param pluginExtension the extension used to determine the plugin
      * @param name defines the name of the file, if the data is binary
-     * @returns The loaded plugin
+     * @deprecated Please use ImportMeshAsync instead
```
Comment (ryantrem) [inline packages/dev/core/src/Loading/sceneLoader.ts:1550]: Maybe we should mark the whole class as deprecated and point people to the module level functions, and remove @experimental from them now?

**Review by sebavan — APPROVED** [review]

File: packages/dev/core/src/Misc/webRequest.fetch.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { WebRequest } from "./webRequest";
+
+/**
+ * Fetches a resource from the network
+ * @param url defines the url to fetch the resource from
+ * @param options defines the options to use when fetching the resource
+ * @returns a promise that resolves when the resource is fetched
+ * @internal
+ */
+export function _FetchAsync(
```
Comment (ryantrem) [inline packages/dev/core/src/Misc/webRequest.fetch.ts:10]: Seems like this is kind of like a fetch polyfill. Couldn't we actually just use `fetch` directly and use an existing `fetch` polyfill that is built on top of `XmlHttpRequest` in cases where we need it (e.g. Babylon Native)? @bghgary

File: packages/dev/core/src/Loading/sceneLoader.ts (hunk 3 above)
Comment (deltakosh) [reply packages/dev/core/src/Loading/sceneLoader.ts:1550]: I do not want to break the entire classes.

File: packages/dev/core/src/Misc/webRequest.fetch.ts (hunk 4 above)
Comment (deltakosh) [reply packages/dev/core/src/Misc/webRequest.fetch.ts:10]: Agreed. And it is internal so we can remove it when native will have fetch support

File: packages/dev/core/src/Loading/sceneLoader.ts (hunk 2 above)
Comment (deltakosh) [reply packages/dev/core/src/Loading/sceneLoader.ts:807]: Function is flagged with asynchronous keyword now hence the promise

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by deltakosh** [comment]
actually impact is bigger than anticipated. Closing for now

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15798/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15798/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15798/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15798/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15798/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15798/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15798/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15798/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15798/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15798/merge/testResults/webgpuplaywright/index.html

**Review by sebavan — APPROVED** [review]

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15798/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15798/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by deltakosh** [comment]
Waiting for the vote here to finish to decide if we merge or not: https://forum.babylonjs.com/t/potential-new-breaking-change-please-chime-in/54651/1

**Pull request merged**